quickwit-indexing/src/source/push_api_source.rs

```diff
@@ -0,0 +1,355 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::time::Duration;
+
+use async_trait::async_trait;
+use quickwit_actors::{ActorExitStatus, Mailbox};
+use quickwit_config::PushApiSourceParams;
+use quickwit_metastore::checkpoint::{CheckpointDelta, PartitionId, Position, SourceCheckpoint};
+use quickwit_proto::push_api::{
+    CreateQueueRequest, FetchRequest, FetchResponse, QueueExistsRequest,
+};
+use quickwit_pushapi::{get_push_api_service, iter_doc_payloads, PushApiService};
+use serde::Serialize;
+
+use super::file_source::BATCH_NUM_BYTES_THRESHOLD;
+use super::{Source, SourceContext, TypedSourceFactory};
+use crate::actors::Indexer;
+use crate::models::RawDocBatch;
+
+#[derive(Default, Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct PushApiSourceCounters {
+    pub previous_offset: u64,
+    pub current_offset: u64,
+    pub num_docs_processed: u64,
+}
+
+pub struct PushApiSource {
+    params: PushApiSourceParams,
+    push_api_mailbox: Mailbox<PushApiService>,
+    counters: PushApiSourceCounters,
+}
+
+impl PushApiSource {
+    pub async fn make(
+        params: PushApiSourceParams,
+        push_api_mailbox: Mailbox<PushApiService>,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let partition_id = PartitionId::from(params.index_id.clone());
+        let offset = if let Some(Position::Offset(offset_str)) =
+            checkpoint.position_for_partition(&partition_id).cloned()
+        {
+            offset_str.parse::<u64>()?
+        } else {
+            0
+        };
+
+        ensure_queue_exist_for_index(&push_api_mailbox, params.index_id.clone()).await?;
```

Comment:
Why do we need to ensure that the queue exists here?

Reply:
I think it is better to not create the queue here.
I am not sure it is helping in any way, and it makes the code more error prone.

By leaving the creation to the push side, we end up creating it in only one place of the code, which makes your mental model simpler.
It avoid the current race condition: there exists a timing for which `ensure_queue_exists_for_index` returns an error, just because someone managed to push a message in the middle of this function call, hence creating the queue right before.

Reply:
Ok